File: libraries/python/getpatter/engines/openai.py
"""OpenAI Realtime engine marker for Patter."""

from __future__ import annotations

import os
from dataclasses import dataclass

__all__ = ["Realtime"]


@dataclass(frozen=True)
class Realtime:
    """OpenAI Realtime API engine config.

    Holds the minimal settings needed by the Patter server to instantiate
    :class:`getpatter.providers.openai_realtime.OpenAIRealtimeAdapter` at call time.

    Example::

        from getpatter.engines import openai

        engine = openai.Realtime()                     # reads OPENAI_API_KEY
        engine = openai.Realtime(voice="nova", model="gpt-4o-mini-realtime-preview")
    """

    api_key: str = ""
    voice: str = "alloy"
    model: str = "gpt-4o-mini-realtime-preview"

    def __post_init__(self) -> None:
        key = self.api_key or os.environ.get("OPENAI_API_KEY", "")
        if not key:
            raise ValueError(
                "OpenAI Realtime engine requires an api_key. Pass "
                "api_key='sk-...' or set OPENAI_API_KEY in the environment."
            )
        object.__setattr__(self, "api_key", key)

    @property
    def kind(self) -> str:
        """Stable discriminator used for Phase 2 dispatch."""
        return "openai_realtime"

File: libraries/typescript/src/engines/openai.ts
/** OpenAI Realtime engine — marker class for Patter client dispatch. */

/** Constructor options for the OpenAI `Realtime` engine marker. */
export interface RealtimeOptions {
  /** API key. Falls back to OPENAI_API_KEY env var when omitted. */
  apiKey?: string;
  /** Realtime model. Defaults to gpt-4o-mini-realtime-preview. */
  model?: string;
  /** Voice preset. Defaults to alloy. */
  voice?: string;
}

/**
 * OpenAI Realtime engine marker.
 *
 * @example
 * ```ts
 * import * as openai from "getpatter/engines/openai";
 * const engine = new openai.Realtime();                     // reads OPENAI_API_KEY
 * const engine = new openai.Realtime({ voice: "alloy" });
 * ```
 */
export class Realtime {
  readonly kind = "openai_realtime" as const;
  readonly apiKey: string;
  readonly model: string;
  readonly voice: string;

  constructor(opts: RealtimeOptions = {}) {
    const key = opts.apiKey ?? process.env.OPENAI_API_KEY;
    if (!key) {
      throw new Error(
        "OpenAI Realtime requires an apiKey. Pass { apiKey: 'sk-...' } or " +
          "set OPENAI_API_KEY in the environment.",
      );
    }
    this.apiKey = key;
    this.model = opts.model ?? "gpt-4o-mini-realtime-preview";
    this.voice = opts.voice ?? "alloy";
  }
}
